### DISCUSSION TITLE
Implement `navigator.credentials` APIs to be executable on Node.js

### DISCUSSION THREAD

**Discussion by mirismaili** [MasterKale/SimpleWebAuthn#649] [Ideas]
For testing scenarios, I need to perform client-side actions in non-browser environments, such as Node.js.

Another use case is automating the login process on a website (using a passkey) via a script running on Node.js. This script should at least be able to perform the actions `navigator.credentials.create` and `navigator.credentials.get` on Node.js.

However, I couldn't find such functionality anywhere.

**Comment by MasterKale** [comment]
Hello @mirismaili, for something like this I'd recommend stubbing `navigator.credentials.create` and `navigator.credentials.get` to test how they're getting called and to also control what gets returned by them. How you achieve this will be specific to your code base and choice in testing libraries, but here's an example of how I'm achieving this in tests of @simplewebauthn/browser's functionality using Deno's testing library:

https://github.com/MasterKale/SimpleWebAuthn/blob/master/packages/browser/src/methods/startRegistration.test.ts#L74

I used to use Jest for this before the big rewrite recently so if you go back in time a few dozen commits, to before #634, you can see the Jest-specific version of this same test.

I hope this helps.

**Comment by mirismaili** [comment]
Hello @MasterKale, Thanks.

Is this way usable for my second use case too (automating the login process on a website using passkey)?

**Reply by MasterKale** [reply]
For this you could use something like https://pptr.dev to run tests against a page rendered in a headless Chrome/Firefox/etc... You're essentially writing Node code to call Puppeeter which will then perform operations in a browser instance. WebAuthn's Virtual Authenticators can also be set up in browser calls to then let the browser automate the authenticator response:

https://w3c.github.io/webauthn/#sctn-automation-virtual-authenticators
